pkg/backup/compaction_processor.go

```diff
@@ -0,0 +1,465 @@
+	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
+	spec execinfrapb.CompactBackupsSpec,
+	entryCh chan execinfrapb.RestoreSpanEntry,
+	encryption *jobspb.BackupEncryptionOptions,
+	progCh chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress,
+	store cloud.ExternalStorage,
+) error {
+	var fileEncryption *kvpb.FileEncryptionOptions
+	if encryption != nil {
+		fileEncryption = &kvpb.FileEncryptionOptions{Key: encryption.Key}
+	}
+	sinkConf := backupsink.SSTSinkConf{
+		ID:        execCfg.DistSQLSrv.NodeID.SQLInstanceID(),
+		Enc:       fileEncryption,
+		ProgCh:    progCh,
+		Settings:  &execCfg.Settings.SV,
+		ElideMode: spec.ElideMode,
+	}
+	sink, err := backupsink.MakeSSTSinkKeyWriter(sinkConf, store, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := sink.Flush(ctx); err != nil {
+			log.Warningf(ctx, "failed to flush sink: %v", err)
+			logClose(ctx, sink, "SST sink")
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case entry, ok := <-entryCh:
+			if !ok {
+				return nil
+			}
+
+			sstIter, err := openSSTs(ctx, execCfg, entry, fileEncryption, spec.EndTime)
+			if err != nil {
+				return errors.Wrap(err, "opening SSTs")
+			}
+
+			if err := compactSpanEntry(ctx, sstIter, sink); err != nil {
+				return errors.Wrap(err, "compacting span entry")
+			}
+		}
+	}
+}
+
+func compactionChainFromSpec(
+	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
+	spec execinfrapb.CompactBackupsSpec,
+	user username.SQLUsername,
+) (compactionChain, *jobspb.BackupEncryptionOptions, error) {
+	kmsEnv := backupencryption.MakeBackupKMSEnv(
+		execCfg.Settings,
+		&execCfg.ExternalIODirConfig,
+		execCfg.InternalDB,
+		user,
+	)
+	prevManifests, localityInfo, encryption, allIters, err := getBackupChain(
+		ctx, execCfg, user, spec.Destination, spec.Encryption, spec.EndTime, &kmsEnv,
+	)
+	if err != nil {
+		return compactionChain{}, nil, err
+	}
+	compactChain, err := newCompactionChain(
+		prevManifests, spec.StartTime, spec.EndTime, localityInfo, allIters,
+	)
+	if err != nil {
+		return compactionChain{}, nil, err
+	}
+	return compactChain, encryption, nil
+}
+
+func genIntersectingSpansForCompaction(
+	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
+	spec execinfrapb.CompactBackupsSpec,
+	compactChain compactionChain,
+	intersectSpanCh chan execinfrapb.RestoreSpanEntry,
+) error {
+	backupLocalityMap, err := makeBackupLocalityMap(compactChain.compactedLocalityInfo, spec.User())
+	if err != nil {
+		return err
+	}
+
+	introducedSpanFrontier, err := createIntroducedSpanFrontier(
+		compactChain.backupChain, compactChain.lastBackup().EndTime,
+	)
+	if err != nil {
+		return err
+	}
+	defer introducedSpanFrontier.Release()
+
+	targetSize := targetRestoreSpanSize.Get(&execCfg.Settings.SV)
+	maxFiles := maxFileCount.Get(&execCfg.Settings.SV)
+
+	var fsc fileSpanComparator = &exclusiveEndKeyComparator{}
+	filter, err := makeSpanCoveringFilter(
+		spec.Spans,
```

Comment:
A slight difference here is that in planning, the span covering filter is generated from `manifest.Spans`, whereas `spec.Spans` has had the completed spans removed. The tests didn't seem to have an issue with this, although not sure if that is a problem?

Reply:
we should use the same spans for planning and here. spec.Spans only contains the spans assigned to the proc.

Reply:
Oh actually `spec.Spans` is the set of all spans that need to be compacted, excluding any that have already been completed (it's what is passed to `generateAndSendImportSpans` in both planning and the processor). `AssignedSpans` is what is assigned to the proc.